### DISCUSSION TITLE
Arana 分布式事务设计

### DISCUSSION THREAD

**Discussion by luky116** [arana-db/arana#822] [Ideas]
# 核心流程
1、正常流程：用户提交事务流程如下：
![](https://cdn.nlark.com/yuque/__puml/8f2765bc11e260cc151fa381debccc4d.svg#lake_card_v2=eyJ0eXBlIjoicHVtbCIsImNvZGUiOiJAc3RhcnR1bWxcblxuYXV0b251bWJlclxuXG5hY3RvciBcIueUqOaIt1wiIGFzIFVzZXJcbnBhcnRpY2lwYW50IFwiQXJhbmFcIiBhcyBBcmFuYVxucGFydGljaXBhbnQgXCJEQjFcIiBhcyBEQjFcbnBhcnRpY2lwYW50IFwiREIyXCIgYXMgREIyXG5cblxuYWN0aXZhdGUgVXNlclxuXG5Vc2VyIC0-IEFyYW5hOiDosIPnlKggYmVnaW4g5pa55rOV5byA5ZCv5YWo5bGA5LqL5YqhXG5hY3RpdmF0ZSBBcmFuYVxuXG5BcmFuYSAtPiBBcmFuYTog5YWo5bGA5LqL5Yqh6KGo5o-S5YWl5LiA5p2h5YWo5bGA5LqL5Yqh6K6w5b2V77yM54q25oCB5Li64oCcc3RhcnRlZOKAnVxuXG5BcmFuYSAtPiBVc2VyOiDov5Tlm57lhajlsYDkuovliqFJRFxuXG5hY3RpdmF0ZSBEQjFcblVzZXIgLT4gQXJhbmE6IOaJp-ihjOS4muWKoVNRTDFcblxuYWx0IOesrOS4gOasoeaJp-ihjOS6i-WKoVNRTFxuICAgIEFyYW5hIC0-IERCMTogWEEgU3RhcnRcblx0XHRBcmFuYSAtPiBEQjE6IOaJp-ihjOS4muWKoVNRTFxuZWxzZSDpnZ7nrKzkuIDmrKHmiafooYzkuovliqFTUUxcblx0XHRBcmFuYSAtPiBEQjE6IOaJp-ihjOS4muWKoVNRTFxuZW5kXG5cbkRCMSAtPiBBcmFuYTog6L-U5Zue5omn6KGM57uT5p6cXG5cbmRlYWN0aXZhdGUgREIxXG5cbmFsdCBEQjHmiafooYzmiJDlip9cbiAgICBBcmFuYSAtPiBBcmFuYTog5b6A5YiG5pSv5LqL5Yqh6KGo5o-S5YWl5LiA5p2h5a2Q5LqL5Yqh6K6w5b2V77yMXFxuIOeKtuaAgeS4uuKAnHN0YXJ0ZWTigJ1cblx0XHRBcmFuYSAtPiBVc2VyOiDov5Tlm57lrZDkuovliqHmiafooYzmiJDlip9cbmVsc2UgREIx5omn6KGM5aSx6LSlXG5cdFx0QXJhbmEgLT4gQXJhbmE6IOW-gOWIhuaUr-S6i-WKoeihqOaPkuWFpeS4gOadoeWtkOS6i-WKoeiusOW9le-8jFxcbiDnirbmgIHkuLrigJxmYWlsZWTigJ1cblx0XHRBcmFuYSAtPiBVc2VyOiDov5Tlm57lrZDkuovliqHmiafooYzlpLHotKVcbmVuZFxuXG5hbHQg5rWB56iL5ZKM56ys5LiA5p2hU1FM5LiA5qC377yM55yB55Wl6YOo5YiG5rWB56iLLi4uLlxuICAgIFVzZXIgLT4gQXJhbmE6IOaJp-ihjOS4muWKoVNRTDIuLi4uLi5cbmVuZFxuXG5ub3RlIHJpZ2h0IG9mIFVzZXI6IOi_memHjOWBh-iuvuaJgOacieWtkOS6i-WKoemDveaJp-ihjOaIkOWKn1xuXG5Vc2VyIC0-IEFyYW5hOiDosIPnlKggY29tbWl0IOaWueazle-8iOS8oOmAkuWFqOWxgOS6i-WKoUlE77yJXG5cbkFyYW5hIC0-IEFyYW5hOiDmm7TmlrDlhajlsYDkuovliqHnirbmgIHkuLrigJxjb21taXR0aW5n4oCdXG5cbkFyYW5hIC0-IEFyYW5hOiDmoLnmja7lhajlsYDkuovliqFJRO-8jOafpeaJvuWHuuaJgOacieWtkOS6i-WKoVxuXG5hY3RpdmF0ZSBEQjFcblxuQXJhbmEgLT4gREIxOiBYQSBlbmQgeHh4XG5cbkFyYW5hIC0-IERCMTogWEEgcHJlcGFyZSB4eHhcblxuQXJhbmEgLT4gQXJhbmE6IOWwhuWtkOS6i-WKoTHnirbmgIHnva7kuLrigJxwcmVwYXJlZOKAnVxuXG5hY3RpdmF0ZSBEQjJcblxuQXJhbmEgLT4gREIyOiBYQSBlbmQgeHh4XG5cbkFyYW5hIC0-IERCMjogWEEgcHJlcGFyZSB4eHhcblxuQXJhbmEgLT4gQXJhbmE6IOWwhuWtkOS6i-WKoTLnirbmgIHnva7kuLrigJxwcmVwYXJlZOKAnVxuXG5BcmFuYSAtPiBEQjE6IFhBIGNvbW1pdCB4eHhcblxuQXJhbmEgLT4gQXJhbmE6IOWwhuWtkOS6i-WKoTHnirbmgIHnva7kuLrigJxjb21taXR0ZWTigJ1cblxuZGVhY3RpdmF0ZSBEQjFcblxuQXJhbmEgLT4gREIyOiBYQSBjb21taXQgeHh4XG5cbkFyYW5hIC0-IEFyYW5hOiDlsIblrZDkuovliqEy54q25oCB572u5Li64oCcY29tbWl0dGVk4oCdXG5cbmRlYWN0aXZhdGUgREIyXG5cbkFyYW5hIC0-IEFyYW5hOiDlsIblhajlsYDkuovliqHnirbmgIHnva7kuLrigJxjb21taXR0ZWTigJ1cblxuQXJhbmEgLT4gVXNlcjog6L-U5Zue5LqL5Yqh5o-Q5Lqk5oiQ5YqfXG5cblxuQGVuZHVtbCIsInVybCI6Imh0dHBzOi8vY2RuLm5sYXJrLmNvbS95dXF1ZS9fX3B1bWwvOGYyNzY1YmMxMWUyNjBjYzE1MWZhMzgxZGViY2NjNGQuc3ZnIiwiaWQiOiJ4VUhSdCIsIm1hcmdpbiI6eyJ0b3AiOnRydWUsImJvdHRvbSI6dHJ1ZX0sImNhcmQiOiJkaWFncmFtIn0=)

2、正常流程：用户回滚事务流程如下：
![](https://cdn.nlark.com/yuque/__puml/2c248a1cee9219f30e813715171dc7ee.svg#lake_card_v2=eyJ0eXBlIjoicHVtbCIsImNvZGUiOiJAc3RhcnR1bWxcblxuYXV0b251bWJlclxuXG5hY3RvciBcIueUqOaIt1wiIGFzIFVzZXJcbnBhcnRpY2lwYW50IFwiQXJhbmFcIiBhcyBBcmFuYVxucGFydGljaXBhbnQgXCJEQjFcIiBhcyBEQjFcblxuXG5hY3RpdmF0ZSBVc2VyXG5cblVzZXIgLT4gQXJhbmE6IOiwg-eUqCBiZWdpbiDmlrnms5XlvIDlkK_lhajlsYDkuovliqFcbmFjdGl2YXRlIEFyYW5hXG5cbkFyYW5hIC0-IEFyYW5hOiDlhajlsYDkuovliqHooajmj5LlhaXkuIDmnaHlhajlsYDkuovliqHorrDlvZXvvIznirbmgIHkuLrigJxzdGFydGVk4oCdXG5cbkFyYW5hIC0-IFVzZXI6IOi_lOWbnuWFqOWxgOS6i-WKoUlEXG5cbmFjdGl2YXRlIERCMVxuVXNlciAtPiBBcmFuYTog5omn6KGM5Lia5YqhU1FMMVxuXG5hbHQg56ys5LiA5qyh5omn6KGM5LqL5YqhU1FMXG4gICAgQXJhbmEgLT4gREIxOiBYQSBTdGFydFxuXHRcdEFyYW5hIC0-IERCMTog5omn6KGM5Lia5YqhU1FMXG5lbHNlIOmdnuesrOS4gOasoeaJp-ihjOS6i-WKoVNRTFxuXHRcdEFyYW5hIC0-IERCMTog5omn6KGM5Lia5YqhU1FMXG5lbmRcblxuREIxIC0-IEFyYW5hOiDov5Tlm57miafooYznu5PmnpxcblxuZGVhY3RpdmF0ZSBEQjFcblxuYWx0IERCMeaJp-ihjOaIkOWKn1xuICAgIEFyYW5hIC0-IEFyYW5hOiDlvoDliIbmlK_kuovliqHooajmj5LlhaXkuIDmnaHlrZDkuovliqHorrDlvZXvvIxcXG4g54q25oCB5Li64oCcc3RhcnRlZOKAnVxuXHRcdEFyYW5hIC0-IFVzZXI6IOi_lOWbnuWtkOS6i-WKoeaJp-ihjOaIkOWKn1xuZWxzZSBEQjHmiafooYzlpLHotKVcblx0XHRBcmFuYSAtPiBBcmFuYTog5b6A5YiG5pSv5LqL5Yqh6KGo5o-S5YWl5LiA5p2h5a2Q5LqL5Yqh6K6w5b2V77yMXFxuIOeKtuaAgeS4uuKAnGZhaWxlZOKAnVxuXHRcdEFyYW5hIC0-IFVzZXI6IOi_lOWbnuWtkOS6i-WKoeaJp-ihjOWksei0pVxuZW5kXG5cbmFsdCDmtYHnqIvlkoznrKzkuIDmnaFTUUzkuIDmoLfvvIznnIHnlaXpg6jliIbmtYHnqIsuLi4uXG4gICAgVXNlciAtPiBBcmFuYTog5omn6KGM5Lia5YqhU1FMMi4uLi4uLlxuZW5kXG5cbm5vdGUgcmlnaHQgb2YgVXNlcjog6L-Z6YeM5YGH6K6-5pyJ6YOo5YiG5a2Q5LqL5Yqh5omn6KGM5aSx6LSlXG5cblVzZXIgLT4gQXJhbmE6IOiwg-eUqCByb2xsYmFjayDmlrnms5XvvIjkvKDpgJLlhajlsYDkuovliqFJRO-8iVxuXG5BcmFuYSAtPiBBcmFuYTog5pu05paw5YWo5bGA5LqL5Yqh54q25oCB5Li64oCccm9sbGJhY2tpbmfigJ1cblxuQXJhbmEgLT4gQXJhbmE6IOagueaNruWFqOWxgOS6i-WKoUlE77yM5p-l5om-5Ye65omA5pyJ5a2Q5LqL5YqhXG5cblxubG9vcCDpgY3ljobmiYDmnInnmoTlrZDkuovliqFcblx0YWN0aXZhdGUgREIxXG5cdEFyYW5hIC0-IERCMTogWEEgcm9sbGJhY2sgeHh4XG5cdEFyYW5hIC0-IEFyYW5hOiDlsIblrZDkuovliqEx54q25oCB572u5Li64oCccm9sbGJhY2tlZOKAnVxuZW5kXG5cbmRlYWN0aXZhdGUgREIxXG5cbkFyYW5hIC0-IEFyYW5hOiDlsIblhajlsYDkuovliqHnirbmgIHnva7kuLrigJxyb2xsYmFja2Vk4oCdXG5cbkFyYW5hIC0-IFVzZXI6IOi_lOWbnuS6i-WKoeWbnua7muaIkOWKn1xuXG5AZW5kdW1sIiwidXJsIjoiaHR0cHM6Ly9jZG4ubmxhcmsuY29tL3l1cXVlL19fcHVtbC8yYzI0OGExY2VlOTIxOWYzMGU4MTM3MTUxNzFkYzdlZS5zdmciLCJpZCI6Imh4elVHIiwibWFyZ2luIjp7InRvcCI6dHJ1ZSwiYm90dG9tIjp0cnVlfSwiY2FyZCI6ImRpYWdyYW0ifQ==)

3、异常流程：事务处于中间状态
如果 Arana 和 DB 通信中断，导致无法通知 DB 侧将事务及时回滚/提交，使事务一直处于中间状态，需要 Arana 对这些事务手动进行处理。Arana 中有一个 “Agent”定时任务，会定期扫描所有中间状态的事务，进行处理。
![](https://cdn.nlark.com/yuque/__puml/5cbfd9caeeddd6df4f38d145c06b2c70.svg#lake_card_v2=eyJ0eXBlIjoicHVtbCIsImNvZGUiOiJAc3RhcnR1bWxcblxuYXV0b251bWJlclxuXG5cbnBhcnRpY2lwYW50IFwiQXJhbmEgQWdlbnRcIiBhcyBBcmFuYVxucGFydGljaXBhbnQgXCJEQjFcIiBhcyBEQjFcblxuXG5hY3RpdmF0ZSBBcmFuYVxuXG5BcmFuYSAtPiBBcmFuYTog5p-l5om-5Lit6Ze054q25oCB44CB6LaF5pe255qE5YWo5bGA5LqL5YqhIFxcbiAo6Z2eY29tbWl0dGVk5oiWcm9sbGJhY2tlZOeKtuaAgSlcblxubG9vcCDpgY3ljobmiYDmnInnmoTlhajlsYDkuovliqFcblxuQXJhbmEgLT4gQXJhbmE6IOaJvuWHuuWFqOWxgOS6i-WKoeS4i-aJgOacieWkhOS6juS4remXtOeKtuaAgeeahOWtkOS6i-WKoSBcXG4gKOmdnmNvbW1pdHRlZOOAgXJvbGxiYWNrZWTmiJZmYWlsZWTnirbmgIEpXG5cblxubG9vcCDpgY3ljobmiYDmnInnmoTlrZDkuovliqFcblx0YWN0aXZhdGUgREIxXG5cdEFyYW5hIC0-IERCMTogWEEgUkVDT1ZFUiB4eHhcblx0REIxIC0-IEFyYW5hOiDov5Tlm57kuovliqHnmoTnirbmgIFcblx0QXJhbmEgLT4gREIxOiDlj5HpgIEgWEEgUm9sbGJhY2sgeHh4XG5cdEFyYW5hIC0-IEFyYW5hOiDmm7TmlrDlrZDkuovliqHnirbmgIHkuLrigJxyb2xsYmFja2Vk4oCdXG5lbmRcblxuZGVhY3RpdmF0ZSBEQjFcblxuQXJhbmEgLT4gQXJhbmE6IOabtOaWsOWFqOWxgOS6i-WKoeeKtuaAgeS4uuKAnFJvbGxiYWNrZWTigJ1cblxuZW5kXG5cbkBlbmR1bWwiLCJ1cmwiOiJodHRwczovL2Nkbi5ubGFyay5jb20veXVxdWUvX19wdW1sLzVjYmZkOWNhZWVkZGQ2ZGY0ZjM4ZDE0NWMwNmIyYzcwLnN2ZyIsImlkIjoiSzlDTWEiLCJtYXJnaW4iOnsidG9wIjp0cnVlLCJib3R0b20iOnRydWV9LCJjYXJkIjoiZGlhZ3JhbSJ9)

4、死锁检测
// TODO
# 细节设计
1、Arana 全局事务表 __arana_global_trx_log：
```sql
CREATE TABLE IF NOT EXISTS __arana_global_trx_log
(
  log_id      			bigint(20) auto_increment COMMENT 'primary key',
  txr_id            varchar(255)     NOT NULL COMMENT 'transaction uniq id',
  tenant      			varchar(255)     NOT NULL COMMENT 'tenant info',
  server_id   			int(10) UNSIGNED NOT NULL COMMENT 'arana server node id',
  status      			int(10)          NOT NULL COMMENT '全局事务状态：started,committing,committed,rollbacking,rollbacked',
  expected_end_time datetime        NOT NULL DEFAULT CURRENT_TIMESTAMP "事务期望结束的时间，如果超时未结束，将会由 Agent 进行处理",
  start_time  			datetime        NOT NULL DEFAULT CURRENT_TIMESTAMP,
  update_time 			datetime        NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (log_id),
  UNIQUE KEY (txr_id)
) ENGINE = InnoDB CHARSET = utf8
```
2、Arana 分支事务表 __arana_branch_trx_log：
```sql
CREATE TABLE IF NOT EXISTS __arana_branch_trx_log
(
  log_id      			bigint(20) auto_increment COMMENT 'primary key',
  txr_id     				varchar(255)     NOT NULL COMMENT '全局事务ID',
  branch_id     		varchar(255)     NOT NULL COMMENT '分支事务ID',
  participant_id   	int(10) UNSIGNED NOT NULL COMMENT '参与子事务执行的DB服务ID',
  status      			int(10)          NOT NULL COMMENT '分支事务状态：started,prepared,committed,rollbacked,failed',
  start_time  			datetime        NOT NULL DEFAULT CURRENT_TIMESTAMP,
  update_time 			datetime        NOT NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP,
  PRIMARY KEY (log_id),
  UNIQUE KEY (branch_id)
) ENGINE = InnoDB CHARSET = utf8
```
3、Arana 的 Agent 本质上是一个定时任务，会定期扫描 __arana_global_trx_log 表，查找 expected_end_time 超时，并且处于非“committed”或“rollbacked”状态的全局事务，主动触发分支事务（非“committed”、“rollbacked”或“failed”）进行结束。
4、注意，__arana_global_trx_log 表和 __arana_branch_trx_log 表是属于 Arana 的系统表。
# 参考资料
[1] 这才是真正的分布式事务 XA 详解
[https://juejin.cn/post/7096781044010991629](https://juejin.cn/post/7096781044010991629) 
[2] 分布式系统事务处理：二阶段提交与三阶段提交协议解析
[https://baijiahao.baidu.com/s?id=1781249550136480595&wfr=spider&for=pc](https://baijiahao.baidu.com/s?id=1781249550136480595&wfr=spider&for=pc)
[3] 6 张图带你彻底搞懂分布式事务 XA 模式
[https://developer.aliyun.com/article/783796](https://developer.aliyun.com/article/783796)
[4] 腾讯云TDSQL——分布式事务实现机制
[https://www.modb.pro/db/109733](https://www.modb.pro/db/109733)
[5] Mysql的分布式(XA)真面目
[https://blog.csdn.net/u013487071/article/details/123718874](https://blog.csdn.net/u013487071/article/details/123718874)
[6] TDSQL在分布式事务阶段遇到死锁时如何处理的
[https://www.cnblogs.com/tencentdb/p/15220582.html](https://www.cnblogs.com/tencentdb/p/15220582.html)
[7] 13.3.8 XA事务
[https://zhuanlan.zhihu.com/p/400930855](https://zhuanlan.zhihu.com/p/400930855)
[8] MySQL · 特性分析 · 浅谈 MySQL 5.7 XA 事务改进
[http://mysql.taobao.org/monthly/2017/09/05/](http://mysql.taobao.org/monthly/2017/09/05/)

**Comment by dongzl** [comment]
timestamp 字段类型调整为 datetime

**Reply by luky116** [reply]
done

**Comment by dongzl** [comment]
部署多个 Arana 节点的定时任务，考虑任务互斥情况

**Comment by dongzl** [comment]
https://juejin.cn/post/7295594732948013094
